Compile everything with meson

tobiasdiez opened the pull request:
<!-- ^^^^^
Please provide a concise, informative and self-explanatory title.
Don't put issue numbers in there, do this in the PR body below.
For example, instead of "Fixes #1234" use "Introduce new method to calculate 1+1"
-->
<!-- Describe your changes here in detail -->

<!-- Why is this change required? What problem does it solve? -->
<!-- If this PR resolves an open issue, please link to it here. For example "Fixes #12345". -->
<!-- If your change requires a documentation PR, please link it appropriately. -->

Add meson configuration to compile sagelib with meson. Basic developer docs are added as well (meant as background info, not real installation instructions and thus under "developer").

### :memo: Checklist

<!-- Put an `x` in all the boxes that apply. -->
<!-- If your change requires a documentation PR, please link it appropriately -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!-- Feel free to remove irrelevant items. -->

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [ ] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### :hourglass: Dependencies

<!-- List all open PRs that this PR logically depends on
- #12345: short description why this is a dependency
- #34567: ...
-->

<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

tornaria commented on src/sage/cpython/dict_del_by_value.pyx:
```diff
@@ -48,7 +48,7 @@ cdef extern from "dict_internal.h":
 
 # dk_lookup was removed in python 3.11
-DEF HAS_DK_LOOKUP = PY_VERSION_HEX < 0x30b0000
+DEF HAS_DK_LOOKUP = False
```
This is precisely what I had in mind.

tobiasdiez replied:
Nice. In this case, this PR is actually ready to go.

tobiasdiez committed:
Merge branch 'rel_import_modules' into meson-compile

tobiasdiez committed:
Merge branch 'rel_headers' into meson-compile

tobiasdiez committed:
add version to pyproject.toml

tobiasdiez committed:
fix test command in ci

tobiasdiez committed:
Merge remote-tracking branch 'origin/develop' into meson-compile

tobiasdiez committed:
add more meson build files

tobiasdiez committed:
Merge remote-tracking branch 'tobiasdiez/rel_import_rings' into meson-compile

tobiasdiez committed:
More meson files

tobiasdiez committed:
Merge branch 'rel_import_ntl' into meson-compile

tobiasdiez committed:
Add interpreters

tobiasdiez committed:
Merge branch 'interpreters_outdir' into meson-compile

tobiasdiez committed:
Merge remote-tracking branch 'tobiasdiez/rel_headers' into meson-compile

tobiasdiez committed:
Merge remote-tracking branch 'mkoeppe/absolute_cimport_sage_graphs' into meson-compile

tobiasdiez committed:
Merge remote-tracking branch 'mkoeppe/absolute_cimport_sage_libs' into meson-compile

tobiasdiez committed:
And another round of meson files

tobiasdiez committed:
Fix matrix

tobiasdiez committed:
Merge remote-tracking branch 'mkoeppe/absolute_cimport_sage_geometry' into meson-compile

tobiasdiez committed:
Last set of meson files

tobiasdiez committed:
Last set of relative imports

tobiasdiez committed:
cleanup

tobiasdiez committed:
fix ci

tobiasdiez committed:
Improve symbolic compilation

mkoeppe commented on src/sage/__init__.py:
```diff
@@ -0,0 +1,2 @@
+# This is only here to mark the sage directory as a Python package
```
Can't add this file. We removed it so that `sage` is a PEP 420 implicit namespace package.

tobiasdiez replied:
As far as I can see, that change only worked since you replaced it with sage's own implementation of init files (namely `all`). Happy to hear alternative suggestions. Each init file has a comment explaining why its necessary.

mkoeppe replied:
Well, Cython 3 purports to support PEP 420 implicit namespace packages; see #36228. There may still be some defects in it. In fact, I was keeping my "replace relative cimports" patch in #35095 instead of upstreaming it because I was hoping that better PEP 420 support will arrive later in the Cython 3.0.x series.

tobiasdiez replied:
Sure, once Cython supports implicit namespaces fully, these init files can be deleted.

mkoeppe commented on pyproject.toml:
```diff
@@ -0,0 +1,10 @@
+[build-system]
```
As I have said before in https://github.com/sagemath/sage/pull/36489#discussion_r1375467483, I disagree with making this structural change of making SAGE_ROOT the root of a Python project by introducing the file `pyproject.toml`. 

In https://github.com/sagemath/sage/pull/36503#discussion_r1367947146 you denied that you're even making this structural change, but here on the PR it is now fully formed, by declaring a new Python project ("sagemath"). 

I have made the reason for my objection clear before: Both this file (and the introduction of the file `meson.build` in SAGE_ROOT, as discussed in  https://github.com/sagemath/sage/pull/36489#discussion_r1375152978) mislead users into assuming that they can build Sage using `pip install .` and/or `meson ...` instead of using `./configure && make`. This is not true because the latter are the build system of the Sage **distribution**, whereas you have developed a build system for the (monolithic) Sage **library** -- with is the Python project already declared in `src/pyproject.toml` named "sagemath-standard". 

The correct solution is to edit `src/pyproject.toml` so that the Sage library uses the mesonpy build system instead of setuptools (delete setup.py), and to use `src/meson.build` as the toplevel meson file. This affects the default (monolithic editable) build. Users can use `./configure --disable-editable` or `./configure --disable-editable --enable-wheels` to switch back to a setuptools-based build system.

As I explained in https://github.com/sagemath/sage/pull/36489#issuecomment-1784211893, changing the structure of SAGE_ROOT in this PR is simply not necessary for the goal of the PR -- building the Sage library with meson, which makes the bundling inappropriate.

tobiasdiez replied:
> I have made the reason for my objection clear before: Both this file (and the introduction of the file `meson.build` in SAGE_ROOT, as discussed in [#36489 (comment)](https://github.com/sagemath/sage/pull/36489#discussion_r1375152978)) mislead users into assuming that they can build Sage using `pip install .` and/or `meson ...` instead of using `./configure && make`. This is not true because the latter are the build system of the Sage **distribution**, whereas you have developed a build system for the (monolithic) Sage **library** -- with is the Python project already declared in `src/pyproject.toml` named "sagemath-standard".

I've developed an alternative for make for the sage project. In this meson alternative, there is not such a distinction between sage-the-distribution and sage-library. There is one project, called sage (or sagemath if you prefer). You can build it using `pip install .` (plus some args) or using `meson compile` (plus some args) from the root. The only difference to the current `make build` command is that the developer is responsible for setting up the surrounding build environment (i.e. if dependencies are not present, the build fails). This is very much similar to scipy/numpy/name another meson project.
 
> The correct solution is to edit `src/pyproject.toml` so that the Sage library uses the mesonpy build system instead of setuptools (delete setup.py), and to use `src/meson.build` as the toplevel meson file. This affects the default (monolithic editable) build. Users can use `./configure --disable-editable` or `./configure --disable-editable --enable-wheels` to switch back to a setuptools-based build system.

I don't think the meson-based system can directly function as a new default. It's unrealistic that it will directly as stable as the existing setuptools-based one. So it should be opt-in instead of opt-out. So the root directory is the only option I see to place the new pyprojecttoml file.

mkoeppe replied:
> I've developed an alternative for make for the sage project. In this meson alternative, there is not such a distinction between sage-the-distribution and sage-library.

Try to explain how what you are doing is different from building the Sage library. 

(Edit: Previously asked but not answered in the thread starting at https://github.com/sagemath/sage/pull/36489#discussion_r1367333358)

mkoeppe replied:
> The only difference to the current `make build` command is that the developer is responsible for setting up the surrounding build environmen

Does this not mean that it is doing nothing of what the Sage distribution does in addition to the Sage library?

mkoeppe replied:
> [...] So the root directory is the only option I see to place the new pyprojecttoml file.

Happy to help: 
- Option 1: An environment variable could switch `src/pyproject.toml` from using setuptools to using mesonpy. Trivial to do by declaring a custom PEP-517/660 build api (about 10 lines of code); reference: #29845 
- Option 2: A directory `pkgs/sagemath-standard_meson` with pyproject.toml.m4

tobiasdiez replied:
> > I've developed an alternative for make for the sage project. In this meson alternative, there is not such a distinction between sage-the-distribution and sage-library.
> 
> Try to explain how what you are doing is different from building the Sage library.

Yes of course its building the sage library....that's also the end-goal of sage-the-distribution.

Sage-the-library doesn't have any means to detect and link libraries that are needed for its compilation (except a few runtime fallbacks). Finding and setting these dependencies up is currently the job of sage-the-distribution. This part of sage-the-distribution is covered by meson.

mkoeppe replied:
> Sage-the-library doesn't have any means to detect and link libraries that are needed for its compilation

That's false. It is using pkg-config for this purpose.

tobiasdiez committed:
improve docs

tobiasdiez committed:
git ignore old so files

tobiasdiez committed:
use dependency to find singular

tobiasdiez committed:
remove outdated comment

tobiasdiez committed:
remove gmpxx in symbolic

tobiasdiez committed:
fix a few runtime errors

tobiasdiez committed:
Merge remote-tracking branch 'tobiasdiez/rel_import_modules' into pr/tobiasdiez/36524

tobiasdiez committed:
Properly compile dynamics

tobiasdiez committed:
properly compile sat

tobiasdiez committed:
ci: run tests

tobiasdiez committed:
Fix a few runtime errors

tobiasdiez committed:
Add missing singular dependencies

Co-authored-by: Michael Orlitzky <michael@orlitzky.com>

tobiasdiez committed:
Drop bogus ginac depencency

Sage needs pynac (a fork of ginac), but it doesn't need ginac itself. To avoid potential problems, we drop ginac from the list o dependencies. This requires us to set inc_pynac to "ginac" (the name of the pynac directory) so that the build can find its headers.

Co-authored-by: Michael Orlitzky <michael@orlitzky.com>

tobiasdiez committed:
Fix a few more runtime errors

tobiasdiez committed:
Remove dependency on https://github.com/sagemath/sage/pull/36489

which is blocked for stupid reasons

tobiasdiez committed:
fix farey compilation

tobiasdiez committed:
remove duplicate dependency

tobiasdiez committed:
Fix farey compilation

tobiasdiez committed:
Remove obsolete uncommented code

tobiasdiez committed:
Try to fix farey header compilation

tobiasdiez committed:
try to fix farey compilation in ci

tobiasdiez committed:
Another try to fix farey

tobiasdiez committed:
Merge remote-tracking branch 'origin/develop' into pr/tobiasdiez/36524

tobiasdiez committed:
Don't install farey header

tobiasdiez committed:
Fix farey import error

tobiasdiez committed:
Merge branch 'develop' into meson-compile

tobiasdiez committed:
Remove old warning about sage.gaprc

tobiasdiez committed:
Fix compilation of interpreters

tobiasdiez committed:
Git-ignore all builddirs

tobiasdiez committed:
Include static maxima file

tobiasdiez committed:
Readd a few all imports

tobiasdiez committed:
Recreate missing init files

tobiasdiez committed:
create dependency on generated interpreters

tobiasdiez committed:
Readd interpreter generation in ci

tobiasdiez committed:
And remove some of the init files again

tobiasdiez committed:
Fix interpreters include directory

tobiasdiez committed:
Fix glpk import error

tobiasdiez committed:
Fix arb runtime error

tobiasdiez committed:
Add missing init file

tobiasdiez committed:
Fix runtime errors due to missing glpk

tobiasdiez committed:
Don't remove init file in interpreters

tobiasdiez committed:
Fix missing pari dependency

tobiasdiez committed:
cleanup ci

tobiasdiez committed:
install ext_data

tobiasdiez committed:
Don't fail fast in ci

tobiasdiez committed:
set env variables correctly

tobiasdiez committed:
Add a few missing headers

tobiasdiez committed:
correctly specify version

tobiasdiez committed:
Properly set SAGE_SHARE to datadir, and warn if cremona is not installed

tobiasdiez committed:
Install more headers

tobiasdiez committed:
Readd init files

tobiasdiez committed:
Fix tests

tobiasdiez committed:
Fix linter

tobiasdiez committed:
Merge branch 'develop' into init_files

tobiasdiez committed:
Merge branch 'develop' into meson-compile

tobiasdiez committed:
Merge remote-tracking branch 'upstream/develop' into meson-compile

tobiasdiez committed:
Merge branch 'meson-compile' of https://github.com/tobiasdiez/sage into meson-compile

tobiasdiez committed:
update meson files

tobiasdiez committed:
fix bernmm compilation

/opt/conda/envs/sage-dev/x86_64-conda-linux-gnu/include/c++/12.3.0/bits/stl_tree.h:772:15: error: static assertion failed: comparison object must be invocable as const
  772 |               is_invocable_v<const _Compare&, const _Key&, const _Key&>,
  /opt/conda/envs/sage-dev/x86_64-conda-linux-gnu/include/c++/12.3.0/bits/stl_tree.h:772:15: note: 'std::is_invocable_v<const bernmm::Item_cmp&, bernmm::Item* const&, bernmm::Item* const&>' evaluates to false

tobiasdiez committed:
fix missing python file

tobiasdiez committed:
fix a few more failing tests by installing necessary files

tobiasdiez committed:
embed file position into cython docstrings

tobiasdiez committed:
Merge remote-tracking branch 'tobiasdiez/init_files' into meson-compile

tobiasdiez committed:
Merge branch 'develop' into init_files

tobiasdiez committed:
Merge branch 'develop' into meson-compile

tobiasdiez committed:
Merge remote-tracking branch 'origin/develop' into pr/tobiasdiez/36524

tobiasdiez committed:
Fix build error

tobiasdiez committed:
cleanup

tobiasdiez committed:
Remove explicit file names from docs

Reason: Cython uses the working dir to relativize the file name (see https://github.com/cython/cython/blob/2f93a5f92a80352e72f5fa9679fd282cb7a497ad/Cython/Compiler/Scanning.py#L199-L200), but meson runs the compilation from a temporary folder (https://github.com/mesonbuild/meson/blob/e991c4d45460c09e5eaee6cb1041edb95d1e990e/mesonbuild/compilers/compilers.py#L832-L879). Hence this the file name relatization is not working.

tobiasdiez committed:
Merge remote-tracking branch 'upstream/develop' into meson-compile

tobiasdiez committed:
Update meson file of sets

tobiasdiez committed:
Revert "Remove old warning about sage.gaprc"

This reverts commit 395c310fbc0c4482dfe05c0ee5915431fa21160f.

tobiasdiez committed:
fix farey header import

tobiasdiez committed:
Print more infos during ci

tobiasdiez committed:
Implement gap root discovery in meson

tobiasdiez committed:
Install missing directory

tobiasdiez committed:
fix failing test

tobiasdiez committed:
docs: uninstall other sage packages

tobiasdiez committed:
include missing cython source file

tobiasdiez committed:
Install missing file in groups

tobiasdiez committed:
Merge remote-tracking branch 'upstream/develop' into init_files

mkoeppe commented on pyproject.toml:
```diff
@@ -0,0 +1,10 @@
+dynamic = ["version"]
+
+[tool.setuptools.dynamic]
+version = {file = ["VERSION.txt"]}
```
```suggestion
```

mkoeppe commented on generate-meson.py:
```diff
@@ -0,0 +1,198 @@
+                    try:
+                        libraries.remove('CYGWIN_SQLITE3_LIBS')
+                    except ValueError:
+                        pass
```
```suggestion
```

mkoeppe commented on generate-meson.py:
```diff
@@ -0,0 +1,198 @@
+# Small script that generates a meson.build file in the given folder.
```
I think this should become a module of `sage_setup`: Like `sage_setup.autogen`, in generates source code.

tobiasdiez replied:
This script is no longer needed and helpful once the initial set of meson files is generated. I'll remove it once the PR has been (or is close to be) positively reviewed.

mkoeppe replied:
I disagree. 

This would put the unnecessary burden on developers to maintain the `meson.build` files.

orlitzky replied:
The auto-generated files are missing a lot of important stuff. it was convenient to have them generate the initial boilerplate, but in the future, we'll be changing the build files one line at a time.

mkoeppe replied:
What is missing that cannot be autogenerated?

tobiasdiez committed:
Merge branch 'develop' into meson-compile

tobiasdiez committed:
Merge branch 'develop' into init_files

tobiasdiez committed:
Remove unused setuptools versioning in pyproject.toml

tobiasdiez committed:
Fix bug with Cython source code not appearing in tracebacks

tobiasdiez committed:
Fix runtime problem with lie algebras

tobiasdiez committed:
Cleanup meson files

tobiasdiez committed:
Update documentation

tobiasdiez committed:
Add meson as an installation option in table of contents

tobiasdiez committed:
Use ccache action

tobiasdiez committed:
Add ccache to project compilers as workaround

tobiasdiez committed:
Remove ccache from conda env

tobiasdiez committed:
Update ccache-action to use main branch

tobiasdiez committed:
Explicitly set path to ccache

tobiasdiez committed:
Fix ccache-action version

tobiasdiez committed:
Update build configuration to force use ccache

tobiasdiez committed:
Fix reference to section-prereqs in meson.rst

tobiasdiez committed:
Fix broken link in installation instructions

tobiasdiez committed:
Fix ccache restore key

tobiasdiez reviewed (commented):
> > The modularized distro's could easily not distribute/install the init files and everything is well.
> That's an oversimplification because in our default, editable install, we run the Sage library directly out of `SAGE_ROOT/src/`. 

meson-python implements a custom loader for editable installs that respects the install plan of meson and hence, even in an editable install, `sage.libs` is a namespace package:
```
>>> import sage.libs
>>> sage.libs.__file__ is None
True
>>> sage.libs.__loader__
<_frozen_importlib_external.NamespaceLoader object at 0x7f85e3807650>
```

tobiasdiez committed:
Merge branch 'develop' into init_files

tobiasdiez committed:
Merge branch 'develop' into meson-compile

tobiasdiez committed:
Update flint meson file

tobiasdiez committed:
Update test command to use 4 parallel processes

tobiasdiez committed:
Merge branch 'develop' into meson-compile

tobiasdiez committed:
Merge branch 'develop' into init_files

tobiasdiez committed:
Merge branch 'develop' into init_files

tobiasdiez committed:
Merge branch 'develop' into meson-compile

tobiasdiez committed:
Remove gap_packages.py from py.install_sources()

tobiasdiez committed:
Add new Python files to various modules

tobiasdiez committed:
Add order_ideal.py to number_field module

tobiasdiez committed:
Delete unused meson files in algebras and combinatorics

tobiasdiez reviewed (commented):
> > If it's included in the diff

> Just search for `/dev/null` in the diff to see the files 

Done noow. Next time please give concete suggestions with all necessary details when reviewing a PR (i.e. a list of files in this case). Thanks!

tobiasdiez committed:
Merge remote-tracking branch 'upstream/develop' into meson-compile

tobiasdiez committed:
Add new files to categories and designs

tobiasdiez committed:
Remove unnecessary meson_version field from project configuration

tobiasdiez committed:
Update install_sources in meson files

tobiasdiez committed:
update cython files

tobiasdiez committed:
Refactor matrix build dependencies and add meataxe dependency for matrix_gfpn_dense

tobiasdiez committed:
Merge branch 'develop' into init_files

tobiasdiez committed:
Try to fix doctest on macos

tobiasdiez committed:
Update library dependencies handling in meson.build files

tobiasdiez committed:
Revert "Try to fix doctest on macos"

This reverts commit 57b8a9ece58cd9ae6256521f1ac9beeb89cf1e5c.

tobiasdiez committed:
Fix exponent overflow bug in RingElement class

tobiasdiez committed:
Improve handling of dependencies

tobiasdiez committed:
Merge branch 'develop' into init_files

tobiasdiez committed:
Merge branch 'develop' into meson-compile

tobiasdiez committed:
Add script for updating Python sources and extension data in meson build files

tobiasdiez committed:
Remove unnecessary code for finding include path for Cython modules

tobiasdiez committed:
Improve code for generating interpreters

tobiasdiez committed:
Update CI workflow to use lock environment file

tobiasdiez committed:
Remove unnecessary arb include directory in meson.build files

tobiasdiez committed:
add meson to conda lock file

tobiasdiez committed:
Remove unnecessary interpreters include_directories in meson.build files

tobiasdiez committed:
Merge branch 'pr/antonio-rojas/37484' into meson-compile

tobiasdiez committed:
Fix compilation (arb)

tobiasdiez committed:
try to fix interpreter build

tobiasdiez committed:
More fun with interpreters

orlitzky commented on src/meson.build:
```diff
@@ -60,10 +60,28 @@ pari = cc.find_library('pari')
+# flint = dependency('flint', version : '>=3.0.0')
+flint = dependency('flint', version : '>=3.0.0', required: false, allow_fallback: false)
+if not flint.found()
+    flint = subproject('flint').get_variable('flint_dep')
```
I don't think this is right.. should it just be falling back to `cc.find_library('flint')` if the `dependency()` call failed?

Currently it's trying to download flint on a machine where it's already installed...

tobiasdiez replied:
If you have flint 2.* installed, the compilation will not succeed as we recently broke backwards compatiblity. So in order to compile sage on recentish systems, I had to add this fallback (see https://github.com/sagemath/sage/pull/37077 for the tests). (Compilation works, but there are now runtime errors related to arb/flint - haven't found the time to investigate this further.)

orlitzky replied:
Nevermind, I assumed that I had a working pkg-config file for my flint-3.0.1, but it looks like it isn't installed by default. One more thing to investigate.

orlitzky replied:
For the record: https://github.com/flintlib/flint/commit/c37bb5b000b2988314134c57e27b9d6423434a6e

tobiasdiez replied:
The pkg-config of flint 3.0 is also not working properly, you need https://github.com/flintlib/flint/pull/1647.

orlitzky replied:
Ah, thanks, but it's installing the `*.pc` file to the wrong location anyway.

tobiasdiez committed:
Add build directory as include directory for Cython compilation

tobiasdiez committed:
Merge branch 'develop' into meson-compile

tobiasdiez committed:
Don't run pytest on builddir

tobiasdiez committed:
Restore --config-settings editable_mode=compat in ci-conda

tobiasdiez committed:
Cleanup workflow

tobiasdiez committed:
Merge remote-tracking branch 'upstream/develop' into init_files

tobiasdiez committed:
Ignore module-not-measured error during code coverage collection

tobiasdiez committed:
Remove acidentially commited submodules

tobiasdiez committed:
Move compilers to src/meson.build

tobiasdiez committed:
Remove acidentially commited submodules

tobiasdiez committed:
Merge branch 'init_files' into meson-compile

tobiasdiez committed:
Fix tests

tobiasdiez committed:
Try installing the generated init file to fix cython module name issues

tobiasdiez committed:
Generate init file for interpreters

tobiasdiez committed:
Improve fflas dependency in meson.build

tobiasdiez committed:
Add fallback for gap

tobiasdiez committed:
Add more fallback subprojects

orlitzky commented on subprojects/fflas-ffpack.wrap:
```diff
@@ -0,0 +1,6 @@
+[wrap-file]
```
You're scaring me, this looks a lot like a copy of sage-the-distro but within the new sagelib build system.

dimpase replied:
I don't see why this all has to be a part of sagelib build. Preparing a chroot or something like this for sagelib to use for its dependencies has to be something separate.

dimpase replied:
OTOH I'm OK with tolerating 10 or 20 such `subprojects/`.

orlitzky replied:
My concern is that once again it's impossible to work on sagelib without having to maintain a mini-distro of other people's software. That's 20 subprojects (with pinned versions) whose build systems can now fail as part of the sagelib build, and that need to be updated in sync with changes to sagelib instead of simply changing the version bound in `meson.build`. For subprojects that aren't available on any distro there might be some benefit to it, but all of sage's dependencies are available everywhere, and if (ever) not... we have sage-the-distro for that. Which hopefully somebody else will be maintaining while I work on sagelib for once.

dimpase replied:
Yes, certainly, my concern too. However, there are cases where there is no correct dependency on your OS.
Are these dependencies conditional (it seems to be the case for GAP, but not for the other ones) ?

orlitzky replied:
The idea is to have them downloaded and bundled only if a system copy is not detected, exactly like sage-the-distro. But we know it's not that simple. If you use the bundled flint, for example, then you must also bundle anything that links to flint to avoid linking in multiple versions of the same symbols. The wrap system can't handle that because (same old story) it's not a real package manager and doesn't do dependency resolution and can't build anything not written in its own language.

Package management is simply better handled by conda/nix/guix or the package manager on your system or even sage-the-distro. But most importantly, it's better if _we_ don't have to handle bug reports and patches for 20 subprojects inside of sagelib all over again.

tobiasdiez replied:
> You're scaring me, this looks a lot like a copy of sage-the-distro but within the new sagelib build system.

Sorry, that was not my intention! I don't think that having a few subprojects that provide fallbacks in case the system-version is too old is a huge problem. My hope was that this would be limited to a few libraries (flint, gap, singular?, linbox? what else?). You have more experience with all the inter-dependencies. Do you think this list becomes too long, even if one only builds what is not provided by relatively recent linux distros? The goal was mostly to ease the migration to the meson setup for existing developers that might have not installed all dependencies, and to test it on more platforms (without any subprojects it only compiled on arch). It was definitely not meant as a mini sage-the-distro hell 😄. By the way, you can disable the fallback mechanism completely in meson.

But thanks both of you for your feedback. I'll remove all subprojects from this PR, and will continue my experiments at #37077.

orlitzky replied:
> > You're scaring me, this looks a lot like a copy of sage-the-distro but within the new sagelib build system.
> 
> Sorry, that was not my intention! I don't think that having a few subprojects that provide fallbacks in case the system-version is too old is a huge problem. My hope was that this would be limited to a few libraries (flint, gap, singular?, linbox? what else?). You have more experience with all the inter-dependencies. Do you think this list becomes too long, even if one only builds what is not provided by relatively recent linux distros? The goal was mostly to ease the migration to the meson setup for existing developers that might have not installed all dependencies, and to test it on more platforms (without any subprojects it only compiled on arch). It was definitely not meant as a mini sage-the-distro hell 😄. By the way, you can disable the fallback mechanism completely in meson.

And I'm sorry for my strong reaction, but I've been traumatized! I think it might be OK to have subprojects in _some_ cases.

Libraries generally cannot be bundled as subprojects because of the transitive linking issue. If meson accepts the system singular and if that singular links to the system flint-2.x, then building the flint-3.x subproject (because the system flint-2.x is not acceptable) won't suffice. When both are linked into sagelib there will be conflicts. This is why we have the `SAGE_SPKG_DEPCHECK` stuff in `spkg-configure.m4` and AFAIK meson doesn't have a way to handle it.  

Groups of packages with complicated version constraints are also out. Givaro, linbox, and fflas-ffpack are good examples of this. Their versions need to agree exactly, so bundling one while accepting the other two from the system can lead to incompatibilities.

Both of those are just instances of the one thing that I'd truly like to avoid though. Namely, having to deal with build failures in other peoples' packages while working on sagelib. If linbox fails to build as part of the sagelib build process, I don't want that to be my problem. And I don't want to be made responsible for sorting out the version constraints in the subprojects before we can loosen the linbox version constraint in meson.build. If all packages come from the same system, we can assume that they're consistent, and that saves us a lot of trouble.

So in light of all that there aren't a lot of packages that I think would make good subprojects. Actual subprojects of sage could be OK, since we're responsible for those bugs anyway. And any packages that are truly unavailable elsewhere, but I don't think we have any of those as dependencies right now, nor should we ever add them. Packages providing only executables and with uncomplicated dependencies would be the only other class that i can think of right now.

For what it's worth, flint is fixed in https://github.com/flintlib/flint/commit/e84002f3cc9c76a72a7ce1f8b834a02d89bd55c3 and I spent a week working on PALP to get it into Gentoo, so now this can be built "out of the box" on Gentoo as well, without the overlay. We're still missing a few data packages but the build doesn't fail because of it.

tobiasdiez replied:
> And I'm sorry for my strong reaction, but I've been traumatized! 

Thanks for opening up and sharing. And, no worries, we all are here for the same reason :smiley:

Jokes aside, thanks for the input and helping me realize that I went to deep, in fact, far deeper than I ever intended to.

tobiasdiez replied:
Removed these wrap files again.

dimpase committed:
Merge branch 'develop' into meson-compile

dimpase committed:
Merge branch 'develop' into meson-compile

tobiasdiez committed:
Merge branch 'develop' into meson-compile

dimpase committed:
Merge branch 'develop' into meson-compile

tobiasdiez committed:
Merge branch 'develop' into pr/tobiasdiez/36524

tobiasdiez committed:
update meson files

tobiasdiez committed:
Autogenerate init files

tobiasdiez committed:
specify all build dependencies

tobiasdiez committed:
Fix build

tobiasdiez committed:
remove subprojects for now

tobiasdiez committed:
fix conda env

tobiasdiez committed:
fix tests

tobiasdiez committed:
fix macos ci

tobiasdiez committed:
downgrade meson-python

tobiasdiez committed:
fix ci

tobiasdiez committed:
fix jinja error

tobiasdiez committed:
ignore build dir

tobiasdiez committed:
clarify build isolation

tobiasdiez committed:
fix import

tobiasdiez committed:
run build only on linux for now

tobiasdiez committed:
remove non-existing pynac dir

tobiasdiez committed:
try to fix import

tobiasdiez committed:
fix sage.all import

tobiasdiez committed:
fix tests

tobiasdiez committed:
fix autogen iterpreters path

tobiasdiez committed:
fix classic build

tobiasdiez committed:
fix import

tobiasdiez committed:
Revert "fix import"

This reverts commit 2a9cd0ed55a0aad343f5f61e1bec0aa8a92ee232.

tobiasdiez committed:
Merge branch 'develop' into meson-compile

dimpase committed:
Merge branch 'develop' into meson-compile

dimpase committed:
unignore a meson.build file

dimpase committed:
add missing matching.py to the list

tobiasdiez committed:
add runtime dependencies to pyproject.toml

tobiasdiez committed:
improve numpy dependency discovery

tobiasdiez committed:
remove duplicate iml dep

tobiasdiez committed:
remove unused subproject

tobiasdiez committed:
cleanup subprojects

tobiasdiez committed:
check for homfly header

tobiasdiez committed:
specify default c and cpp std

tobiasdiez committed:
explain how to set include and library path manually

tobiasdiez committed:
fix imports in tests

tobiasdiez committed:
Revert "improve numpy dependency discovery"

This reverts commit ae996addad8e18d491aef67782a66a4e29f03526.

tobiasdiez committed:
fix pyproject.toml metadata

tobiasdiez committed:
ignore also build dirs for other python versions

tobiasdiez committed:
reformat meson files

tobiasdiez committed:
improve meson update tool

tobiasdiez committed:
fix compilation

tobiasdiez committed:
exclude main files from testing

tobiasdiez committed:
exclude symlink from vscode search

tobiasdiez committed:
further improvements to flint dpendency handling

tobiasdiez committed:
fix test errors due to missing sage_setup

orlitzky committed:
src/sage/libs/giac/meson.build: build extension only if libgiac is found

The giac package is already marked optional, but it isn't really
optional because the cython extension can't be built without libgiac.
Now, the extension will be skipped if giac is absent. Recent changes
in sagelib have made this feasible.

orlitzky committed:
src/sage/plot/plot3d/meson.build: depend on interpreters

The file parametric_surface.pyx cimports Wrapper_rdf from
sage.ext.interpreters.wrapper_rdf. If the interpreters have not yet
been built, the build can fail here. We split parametric_surface.pyx
from the other plot3d extensions, and add the extra dependency only
for the extension (parametric_surface.pyx) that needs it.

orlitzky committed:
meson.build: add "check: true" for the command that creates __init__.py

Meson warns about this when the "check" keyword is absent,

  WARNING: You should add the boolean check kwarg to the run_command call.
           It currently defaults to false,
           but it will default to true in future releases of meson.
           See also: https://github.com/mesonbuild/meson/issues/9300

and the command should indeed succeed, so we now pass "check: true" to
run_command().

tobiasdiez committed:
fix tests

tobiasdiez committed:
improve dependency handling by using disabler

tobiasdiez committed:
add exception to update-meson helper

tobiasdiez committed:
Improve handling of optional dependencies

tobiasdiez committed:
fix build

tobiasdiez committed:
try to fix doctests of sage_setup

tobiasdiez committed:
try with absolute import

tobiasdiez committed:
try to add interpreters.internal to sage-setup/pyproject.toml

tobiasdiez committed:
fix more autogen tests

dimpase committed:
add -D_XOPEN_SOURCE=600 to the right place

dimpase committed:
use newer action

tobiasdiez committed:
Try to fix interpreters compilation

tobiasdiez committed:
improve docs

dimpase reviewed (commented):
You're removing the _XOPEN ... parameter in meson.build for src/stats/distributions

Are you sure it's not needed? Or it's a result of git-basing off an old branch?

tobiasdiez committed:
Readd D_XOPEN_SOURCE c_arg

tobiasdiez committed:
Fix wonderful rst syntax (maybe)

dimpase reviewed (approved):
LGTM

tobiasdiez committed:
fix documentation

tobiasdiez committed:
Add more optional tags for database_graphs

tobiasdiez committed:
properly copy sage.gaprc

tobiasdiez committed:
fix interpreters dependency by specifying generated headers as source

tobiasdiez committed:
Fix installation of sage.gaprc

tobiasdiez committed:
add all pxd files

dimpase reviewed (approved):
LGTM**2
